library.ml

```diff
@@ -3,64 +3,77 @@ open Printf
 open String
 open List
 
-class enemy=object (self)
-  val mutable name:string=""
-  val mutable health:int=0
-  val mutable dmg:int=0
-  method set_health n=health <- n
-  method set_name n=name <- n
-  method set_dmg n=dmg <- n
+module Enemy = struct
+  type t =
+    { name: string ;
+      health: int ;
+      dmg: int ; }
 
-  method appear:unit=
-    printf "An enemy ";print_string [red] name;printf " appeared!\n";
-    self#battle
-  method battle:unit=
-    if health>0 then begin
-      print_string [red] name;printf " attacks you for ";
-      print_string [red] (string_of_int dmg);printf " damage\n";
+  let empty = { name = "" ; health = 0 ; dmg = 0 ; }
+
+  let rec battle t : t=
+    if t.health > 0 then begin
+      print_string [red] t.name;printf " attacks you for ";
+      print_string [red] (string_of_int t.dmg);printf " damage\n";
       printf "What will you do? (";print_string [yellow] "attack";printf ")\n";
-      let response=read_line() in begin
-        printf "\n";
-        if response="attack" then (
-          health <- health+(-1);
-          printf "You did 1 damage!\n";
-          print_string [red] name;
-          if health=0 then printf " is dead!\n" else printf " has %i health left\n" health
-        ) else printf "Your attack missed\n"
-      end;
-      self#battle
-    end
+      let t =
+        let response=read_line() in begin
+          printf "\n";
+          if response="attack" then (
+            let t = {t with health = t.health - 1} in
+            printf "You did 1 damage!\n";
+            print_string [red] t.name;
+            ( if t.health=0
+              then printf " is dead!\n"
+              else printf " has %i health left\n" t.health );
+            t
+          ) else (
+            printf "Your attack missed\n" ;
+            t
+          )
+        end
+      in battle t
+    end else t
+
+  let appear t : t =
+    printf "An enemy ";print_string [red] t.name;printf " appeared!\n";
+    battle t
 end
 
-class room=object (self)
-  val mutable enemies:enemy list=[]
-  val mutable messages:string list=[]
-  val mutable rooms:room list=[]
-  val mutable name:string=""
-  method name=name
-  method set_name (n:string)=name <- n
-  method add_msg n=messages <- messages@[n]
-  method add_enemy n=enemies <- enemies@[n]
-  method add_room n=rooms <- rooms@[n]
+module Room = struct
+  type t =
+    { enemies: Enemy.t list ;
+      messages: string list ;
+      rooms: t list ;
+      name: string ; }
+  let empty = { enemies = [] ; messages = [] ; rooms = [] ;name = ""; }
+  let add_msg n t = {t with messages = t.messages@[n]}
```

Comment:
Instead of operating on the implicit `self` variable of the object instance, we operate on our state type `t` - and return a new immutable `t` with the necessary updates.